aws/src/main/java/org/apache/iceberg/aws/HttpClientCache.java

```diff
@@ -0,0 +1,206 @@
+    ManagedHttpClient managedClient =
+        clients.computeIfAbsent(
+            clientKey,
+            key -> {
+              LOG.debug("Creating new managed HTTP client for key: {}", key);
+              SdkHttpClient httpClient = clientFactory.get();
+              return new ManagedHttpClient(httpClient, key);
+            });
+    // Return the cached ref-counted wrapper
+    return managedClient.acquire();
+  }
+
+  /**
+   * Release a reference to the HTTP client. When the reference count reaches zero, the client is
+   * closed and removed from the cache.
+   *
+   * @param clientKey the key identifying the client to release
+   */
+  void releaseClient(String clientKey) {
+    ManagedHttpClient managedClient = clients.get(clientKey);
+    if (null != managedClient && managedClient.release()) {
+      clients.remove(clientKey, managedClient);
+    }
+  }
+
+  @VisibleForTesting
+  Map<String, ManagedHttpClient> clients() {
+    return Collections.unmodifiableMap(clients);
+  }
+
+  @VisibleForTesting
+  void clear() {
+    clients.values().forEach(ManagedHttpClient::close);
+    clients.clear();
+  }
+
+  /**
+   * Managed HTTP client wrapper that provides reference counting for lifecycle management. The HTTP
+   * client is closed when the reference count reaches zero.
+   */
+  static class ManagedHttpClient implements SdkHttpClient {
+    private final SdkHttpClient httpClient;
+    private final String clientKey;
+    private final AtomicInteger refCount = new AtomicInteger(0);
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+
+    ManagedHttpClient(SdkHttpClient httpClient, String clientKey) {
+      this.httpClient = httpClient;
+      this.clientKey = clientKey;
+      LOG.debug("Created managed HTTP client: key={}", clientKey);
+    }
+
+    /**
+     * Acquire a reference to the HTTP client, incrementing the reference count.
+     *
+     * @return the ref-counted wrapper client
+     * @throws IllegalStateException if the client has already been closed
+     */
+    ManagedHttpClient acquire() {
+      int count = refCount.incrementAndGet();
+      if (closed.get()) {
+        refCount.decrementAndGet();
+        throw new IllegalStateException("Cannot acquire closed HTTP client: " + clientKey);
+      }
+      LOG.debug("Acquired HTTP client: key={}, refCount={}", clientKey, count);
+      return this;
+    }
+
+    /**
+     * Release a reference to the HTTP client, decrementing the reference count. If the count
+     * reaches zero, the client is closed.
+     *
+     * @return true if the client was closed, false otherwise
+     */
+    boolean release() {
+      if (closed.get()) {
+        LOG.warn("Attempted to release already closed HTTP client: key={}", clientKey);
+        return false;
+      }
+
+      int count = refCount.decrementAndGet();
+      LOG.debug("Released HTTP client: key={}, refCount={}", clientKey, count);
+      if (count == 0) {
+        return closeHttpClient();
```

Comment:
This doesn't appear to be thread-safe.  We're decrementing the value and then doing a check against the ref count, but an increment could have come in an incremented the count, which would then cause a problem because we would close the client on an active ref.

It would be safer to enclose the check and close in a synchronized block on the refCount object.

Reply:
Thanks for the catch. I synchronized the acquire and release methods since refCount is an AtomicInteger.

Reply:
We do expose this in the `refCount()` method, which could be stale, but since it really isn't exposed, I feel this is ok.